### DISCUSSION TITLE
The property upsert does not exist for the type xxx.

### DISCUSSION THREAD

**Discussion by yoshinorihisakawa** [prisma/prisma#20047] [Q&A]
The version of prisma is as follows

`Prisma Client (4.16.2 | library)
`

The schema.prisma is as follows

```prisma
model Document {
  id String @id
  project_id String
  url String?
  title String?
  input_type String
  metadata Json
  created_at DateTime @default(now())
  updated_at DateTime @updatedAt
  DocumentSection DocumentSection[]
  Project Project @relation(fields: [project_id], references: [id])

  @@map("documents"))
}

model DocumentSection {
  id String @id
  document_id String
  project_id String
  content String
  token_count Int?
  metadata Json
  embedding Unsupported("vector")
  document Document @relation(fields: [document_id], references: [id])
  Project Project @relation(fields: [project_id], references: [id])

  @@map("document_sections"))
}
```

I get an error when using the following

<img width="1128" alt="スクリーンショット 2023-07-04 9 50 49" src="https://github.com/prisma/prisma/assets/39504693/d87ebf54-39b3-4105-881b-1977cbf7a8d1">

**Comment by ludralph** [comment]
Hi @yoshinorihisakawa 👋 

Did you generate the Prisma Client by running `npx prisma generate`? Also you may need to restart your code editor as well.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Comment by yoshinorihisakawa** [comment]
We have also implemented the methods you suggested, but they do not resolve the issue.

I have investigated the cause and found that upsert is not created for models that use the Unsupported("vector") type.

For the following models, the upsert was created successfully

```prisma
model DocumentSection2 {
  id String @id
  document_id String
  project_id String
  content String
  token_count Int?
  metadata Json
  document Document @relation(fields: [document_id], references: [id])
  Project Project @relation(fields: [project_id], references: [id])

  @@map("document_sections2"))
}
```

I would like to use the Unsupported("vector") type, please investigate the cause!

**Reply by ludralph** [reply]
Hi @yoshinorihisakawa, From the [documentation](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#remarks-4),  If a model contains a required Unsupported type, `prisma.model.create(..)`, `prisma.model.update(...)` and `prisma.model.upsert(...)` are not available in Prisma Client. You can consider making it optional in order to have access to the `upsert` method.

If this answers your question, it would be great if you could mark this **Discussion as answered** to indicate that it has been resolved.

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by ludralph** [reply]
Hey @yoshinorihisakawa ,

Do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Reply by ludralph** [reply]
Hi @yoshinorihisakawa ,

Do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Comment by yoshinorihisakawa** [comment]
@ludralph 
> You can consider making it optional in order to have access to the upsert method.

How exactly can this be accomplished? Please attach documents etc.

**Reply by ludralph** [reply]
Hi @yoshinorihisakawa, you need to modify your schema as shown below:
```prisma
model Document {
  id String @id
  project_id String
  url String?
  title String?
  input_type String
  metadata Json
  created_at DateTime @default(now())
  updated_at DateTime @updatedAt
  DocumentSection DocumentSection[]
 Project Project @relation(fields: [project_id], references: [id])

  @@map("documents")
}

model DocumentSection {
  id String @id
  document_id String
  project_id String
  content String
  token_count Int?
  metadata Json
  embedding Unsupported("vector")?
  document Document @relation(fields: [document_id], references: [id])
 Project Project @relation(fields: [project_id], references: [id])

  @@map("document_sections")
}
```
Note the `?` at the end of the `embedding` field.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by nurul3101** [reply]
@yoshinorihisakawa Did you had a chance to look at Raphael's response? Please let us know in case you still need any help. If not, please consider marking the discussion as answered/closed.

**Reply by nurul3101** [reply]
We are closing this Discussion due to inactivity. Please open a new Discussion if you have a similar problem 🙏
